internal/crypto/internal_ca.go

```diff
@@ -0,0 +1,160 @@
+	if ca, err := GetCA(certFile, keyFile, serialFile); err == nil {
+		return ca, false, err
+	}
+	ca, err := MakeSelfSignedCA(certFile, keyFile, serialFile, subjectName, expireDays)
+	return ca, true, err
+}
+
+func GetCA(certFile, keyFile, serialFile string) (*internalCA, error) {
+	ca, err := oscrypto.GetCA(certFile, keyFile, serialFile)
+	if err != nil {
+		return nil, err
+	}
+	config := TLSCertificateConfig(*ca.Config)
+	return &internalCA{Config: &config, SerialGenerator: ca.SerialGenerator}, err
+}
+
+func MakeSelfSignedCA(certFile, keyFile, serialFile, subjectName string, expiryDays int) (*internalCA, error) {
+	caConfig, err := makeSelfSignedCAConfig(
+		pkix.Name{CommonName: subjectName},
+		time.Duration(expiryDays)*24*time.Hour,
+	)
+	if err != nil {
+		return nil, err
+	}
+	if err = caConfig.WriteCertConfigFile(certFile, keyFile); err != nil {
+		return nil, err
+	}
+
+	var serialGenerator oscrypto.SerialGenerator
+	if len(serialFile) > 0 {
+		// create / overwrite the serial file with a zero padded hex value (ending in a newline to have a valid file)
+		if err := os.WriteFile(serialFile, []byte("00\n"), 0600); err != nil {
+			return nil, err
+		}
+		serialGenerator, err = oscrypto.NewSerialFileGenerator(serialFile)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		serialGenerator = &oscrypto.RandomSerialGenerator{}
+	}
+
+	config := TLSCertificateConfig(*caConfig)
+	return &internalCA{
+		SerialGenerator: serialGenerator,
+		Config:          &config,
+	}, nil
+}
+
+func makeSelfSignedCAConfig(subject pkix.Name, caLifetime time.Duration) (*oscrypto.TLSCertificateConfig, error) {
+	rootcaPublicKey, rootcaPrivateKey, publicKeyHash, err := NewKeyPairWithHash()
+	if err != nil {
+		return nil, err
+	}
+
+	now := time.Now()
+	rootcaTemplate := &x509.Certificate{
+		Subject: subject,
+
+		SignatureAlgorithm: x509.ECDSAWithSHA256,
+
+		NotBefore: now.Add(-1 * time.Second),
+		NotAfter:  now.Add(caLifetime),
+
+		SerialNumber: randomSerial(),
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+
+		AuthorityKeyId: publicKeyHash,
+		SubjectKeyId:   publicKeyHash,
+	}
+	rootcaCert, err := signCertificate(rootcaTemplate, rootcaPublicKey, rootcaTemplate, rootcaPrivateKey)
+	if err != nil {
+		return nil, err
+	}
+	caConfig := &oscrypto.TLSCertificateConfig{
+		Certs: []*x509.Certificate{rootcaCert},
+		Key:   rootcaPrivateKey,
+	}
+	return caConfig, nil
+}
+
+func (caBackend *internalCA) signCertificate(template *x509.Certificate, requestKey crypto.PublicKey) (*x509.Certificate, error) {
+	// Increment and persist serial
+	serial, err := caBackend.SerialGenerator.Next(template)
+	if err != nil {
+		return nil, err
+	}
+	template.SerialNumber = big.NewInt(serial)
+	return signCertificate(template, requestKey, caBackend.Config.Certs[0], caBackend.Config.Key)
```

Comment:
again the length freaks me out could you expose a method that returns the cert and ensures its length as an option

Reply:
That should happen one level up. At this point (in the backend) we return crypto.x509
I will add it to the frontend